library/Core/src/Juvix/Core/IR/Types/Globals.hs

```diff
@@ -410,26 +459,37 @@ deriving instance
   RawGlobalAll NFData ext primTy primVal =>
   NFData (RawFunClause' ext primTy primVal)
 
-type Signature ty ext primTy primVal = Map.Map GlobalName (SigDef ty ext primTy primVal)
+type Signature ext primTy primVal =
+  Map.Map GlobalName (SigDef ext primTy primVal)
+
+type InnerTCSig ext primTy primVal =
+  StateT (Signature ext primTy primVal)
 
 -- Return type of all type-checking functions.
 -- state monad for global signature
 -- TODO move this somewhere
-type TypeCheck ty ext primTy primVal a =
-  StateT (Signature ty ext primTy primVal) IO a
-
-data SigDef extV extT primTy primVal
+newtype TypeCheck ext primTy primVal m a
```

Comment:
would it be possible to move this somewhere? maybe even merge it with `Typecheck.Env.EnvTypecheck'`?

Reply:
Not sure if it's appropriate to merge it with `EnvTypecheck`. The return type shouldn't return a type that's more than it needs I think. I could move it to the `Env` file?

Reply:
Actually that's a good point. Moving it to `Env` as a distinct type is fine. But maybe it should be renamed to `TypecheckDecl` or something since it's going to be specific to that?

Reply:
Moved to `Env`. As for the name, is it OK if I leave it as is because it's the main type checking function.